Dockerfile.s3

```diff
@@ -0,0 +1,154 @@
+# Set edge bind host so localstack can be reached by other containers
+# set library path and default LocalStack hostname
+ENV USER=localstack
+ENV PYTHONUNBUFFERED=1
+
+# Install the latest version of awslocal globally
+# RUN --mount=type=cache,target=/root/.cache \
+#     pip3 install --upgrade awscli awscli-local requests
+
+
+# builder: Stage which installs the dependencies of LocalStack Community
+FROM base as builder
+ARG TARGETARCH
+
+# Install build dependencies to base
+RUN --mount=type=cache,target=/var/cache/apt \
+    apt-get update && \
+        # Install dependencies to add additional repos
+        apt-get install -y gcc
+
+# upgrade python build tools
+RUN --mount=type=cache,target=/root/.cache \
+    (virtualenv .venv && . .venv/bin/activate && pip3 install --upgrade pip wheel setuptools)
+
+# add files necessary to install all dependencies
+ADD Makefile setup.py setup.cfg pyproject.toml ./
+# add the root package init to invalidate docker layers with version bumps
+ADD localstack/__init__.py localstack/
+# add the localstack start scripts (necessary for the installation of the runtime dependencies, i.e. `pip install -e .`)
+ADD bin/localstack bin/localstack.bat bin/localstack-supervisor bin/
+
+# install dependencies to run the LocalStack Pro runtime and save which ones were installed
+RUN --mount=type=cache,target=/root/.cache \
+    make install-s3
+RUN . .venv/bin/activate && pip3 freeze -l > requirements-runtime.txt
+
+# delete the botocore specs for other services (>80mb)
+# TODO: well now it's compressed and it's much lighter: 20mb maybe not worth it
+RUN find .venv/lib/python3.11/site-packages/botocore/data/ -mindepth 1 -maxdepth 1 -type d -not -name s3 -exec rm -rf '{}' \;
+
+
+
+# final stage: Builds upon base stage and copies resources from builder stages
+FROM base
+COPY --from=builder /opt/code/localstack/.venv /opt/code/localstack/.venv
+
+# add project files necessary to install all dependencies
+ADD Makefile setup.py setup.cfg pyproject.toml ./
+# add the localstack start scripts (necessary for the installation of the runtime dependencies, i.e. `pip install -e .`)
+ADD bin/localstack bin/localstack.bat bin/localstack-supervisor bin/
+
+# add the code as late as possible
+ADD localstack/ localstack/
+
+# delete the API specs stubs for other services
+# delete the providers implementations for other services
+# TODO: remove the SQS ones, it depends on 'is_sqs_queue_url' in service_router.py
+# TODO: remove lambda one, testutil.py depends on it for fixtures
+# TODO: we also need dynamodb, don't remove the API for now, until we find a way for tests...
+# and now we need node for CDK... what a mess omg
+# Maybe we need a bit of an heavier image for tests?
+# RUN find localstack/aws/api/ -mindepth 1 -maxdepth 1 -type d -not -name s3 -not -name sqs -not -name lambda_ -exec rm -rf '{}' \;
+    # \ && \
+    # find localstack/services/ -mindepth 1 -maxdepth 1 -type d -not -name s3 -exec rm -rf '{}' \;
```

Comment:
this was a big back and forth for saving some MB, deleting the generated specs, but the tests need them, and it only adds some MB to the image. For now, we can leave it in.